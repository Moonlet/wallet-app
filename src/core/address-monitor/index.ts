import { Notifications } from '../messaging/notifications/notifications';
import { Blockchain } from '../blockchain/types';
import DeviceInfo from 'react-native-device-info';
import { IWalletsState, IAccountState } from '../../redux/wallets/state';
import { CONFIG } from '../../config';

const url = CONFIG.dataApiUrl + '/notifications/register';
const monitoredBlockchains = [Blockchain.ETHEREUM, Blockchain.ZILLIQA];

// const getDeviceToken = (): Promise<string> =>
//     Platform.OS === 'android' ? Notifications.getToken() : getApnsToken();

export const updateAddressMonitorTokens = (wallets: IWalletsState) => {
    const addresses = {};

    Object.keys(wallets || []).forEach(id => {
        wallets[id].accounts.forEach((account: IAccountState) => {
            if (monitoredBlockchains.indexOf(account.blockchain) !== -1) {
                if (!addresses[account.blockchain]) {
                    addresses[account.blockchain] = [];
                }

                addresses[account.blockchain].push(account.address);
            }
        });
    });

    const requestAddresses = Object.keys(addresses).map(blockchain => ({
        blockchain,
        addresses: addresses[blockchain]
    }));

    const deviceId = DeviceInfo.getUniqueId();
    Notifications.getToken().then(token => {
        const request = {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                deviceId,
                tokenType: 'fcm',
                token,
                addresses: requestAddresses
            })
        };

        fetch(url, request);
    });
};
