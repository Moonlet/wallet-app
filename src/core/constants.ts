import { Blockchain } from './blockchain/types';

const BLOCKCHAIN_COINS = {
    [Blockchain.ETHEREUM]: 'ETH',
    [Blockchain.ZILLIQA]: 'ZIL',
    [Blockchain.COSMOS]: 'ATOM',
    [Blockchain.STELLAR]: 'XLM'
};

export { BLOCKCHAIN_COINS };
