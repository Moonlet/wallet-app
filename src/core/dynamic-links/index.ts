import dynamicLinks from '@react-native-firebase/dynamic-links';
import { URL, URLSearchParams } from 'react-native-url-polyfill';
import { captureException as SentryCaptureException } from '@sentry/react-native';
import { IReduxState } from '../../redux/state';
import { NavigationService } from '../../navigation/navigation-service';
import { getChainId } from '../../redux/preferences/selectors';
import { setSelectedBlockchain } from '../../redux/wallets/actions';
import { Blockchain } from '../blockchain/types';
import { getNrPendingTransactions, getSelectedAccount } from '../../redux/wallets/selectors';
import { ITokenState } from '../../redux/wallets/state';
import { store } from '../../redux/config';

export class DynamicLinksService {
    private onLinkListener = null;
    private state: IReduxState = null;

    private handlers = {
        '/:blockchain/:token/validator/:validatorAddress': params => {
            this.navigateValidatorAddress(params);
        },
        '/:blockchain/:token/stake/:validatorAddress': params => {
            this.stakeValidatorAddress(params);
        },
        '/:blockchain/stake': params => {
            this.blockchainStake(params);
        },
        '/stake': params => {
            // TODO: implement later on
            // console.log('stakeDefault', params);
        }
    };

    private navigateValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;
        const tokenSymbol = String(params?.token)?.toUpperCase();

        if (blockchain && validatorAddress && tokenSymbol) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            const chainId = getChainId(this.state, blockchain);

            const selectedAccount = getSelectedAccount(this.state);

            if (!selectedAccount) {
                SentryCaptureException(
                    new Error(
                        `Dynamic Links - no selected account, no wallet, blockchain: ${blockchain}, tokenSymbol: ${tokenSymbol}, validatorAddress: ${validatorAddress}`
                    )
                );
                return;
            }

            const token: ITokenState = selectedAccount.tokens[chainId][tokenSymbol];

            NavigationService.navigate('Validator', {
                blockchain,
                undefined, // validator
                accountIndex: selectedAccount.index,
                token,
                canPerformAction: !getNrPendingTransactions(this.state), // hasPendingTransactions
                options: {
                    validatorAddress,
                    tokenSymbol
                }
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - navigateValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private stakeValidatorAddress(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;
        const validatorAddress = params?.validatorAddress;

        if (blockchain && validatorAddress) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'SelectStakeAccount',
                    key: 'select-stake-account',
                    params: {
                        validatorId: validatorAddress
                    }
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - stakeValidatorAddress, invalid params: ${params}`)
            );
        }
    }

    private blockchainStake(params: any) {
        const blockchain = String(params?.blockchain)?.toUpperCase() as Blockchain;

        if (blockchain) {
            // make sure `blockchain` is selected
            store.dispatch(setSelectedBlockchain(blockchain));

            NavigationService.navigate('SmartScreen', {
                context: {
                    screen: 'StakeNow',
                    step: 'StakeSelectValidator',
                    key: 'stake-now-select-validator'
                },
                navigationOptions: {
                    title: 'Stake now'
                },
                newFlow: true
            });
        } else {
            SentryCaptureException(
                new Error(`Dynamic Links - blockchainStake, invalid params: ${params}`)
            );
        }
    }

    public async configure(state: IReduxState) {
        this.state = state;

        const dl = dynamicLinks();

        this.onLinkListener = dl.onLink(link => {
            this.handleDynamicLink(link);
        });

        // application is in a background state or has fully quit
        dl.getInitialLink().then(link => {
            this.handleDynamicLink(link);
        });
    }

    // app is in the foreground state
    public async handleDynamicLink(link: any) {
        if (link?.url) {
            this.processUrl(link.url, this.handlers, {
                protocol: ['https:'],
                host: [
                    'moonlet.io',
                    'moonlet.xyz',
                    'moonletwallet.com',
                    'moonlet.app',
                    'moonlet.link'
                ]
            });
        } else {
            // Invalid link url, link:${link}`
        }
    }

    public removeListeners() {
        typeof this.onLinkListener === 'function' && this.onLinkListener();
    }

    private pathPatternToRegexp(pathPattern: string) {
        let exp = '^';

        const pathParts = pathPattern.split('/').filter(Boolean);
        const capturingGroups = [];
        for (const part of pathParts) {
            exp += '\\/';
            if (part.indexOf(':') === 0) {
                exp += '([^\\/]*)';
                capturingGroups.push(part.replace(/^:/, ''));
            } else {
                exp += part;
            }
        }

        exp += '$';
        return {
            regexp: new RegExp(exp),
            capturingGroups
        };
    }

    private parsePath(path: string, pathPatternString: string) {
        const pathPattern = this.pathPatternToRegexp(pathPatternString);
        const match = path.match(pathPattern.regexp);

        if (match) {
            const params = {};
            for (let i = 0; i < pathPattern.capturingGroups.length; i++) {
                params[pathPattern.capturingGroups[i]] = match[i + 1];
            }
            return params;
        }

        return null;
    }

    private processPath(path: string, handlers: any, urlParams = {}) {
        const pathsPatterns = Object.keys(handlers);
        // console.log(path, handlers, urlParams);
        for (const pathPattern of pathsPatterns) {
            if (typeof handlers[pathPattern] === 'function') {
                const params = this.parsePath(path, pathPattern);
                if (params) {
                    return handlers[pathPattern]({
                        ...urlParams,
                        ...params
                    });
                }
            }
        }
    }

    private paramsToObject(entries: any) {
        const result = {};
        for (const [key, value] of entries) {
            // each 'entry' is a [key, value] tupple
            result[key] = value;
        }
        return result;
    }

    private processUrl(url: string, pathHandlers: any = this.handlers, whitelist = {}) {
        const u = new URL(url);

        for (const key of Object.keys(whitelist)) {
            if (u && typeof u[key] === 'string' && whitelist[key].indexOf(u[key]) < 0) {
                SentryCaptureException(
                    new Error(
                        JSON.stringify(
                            `whitelist not matched ${key}=${u[key]} not in whitelist: ${
                                typeof whitelist[key] === 'string'
                                    ? whitelist[key]
                                    : whitelist[key].join(', ')
                            }`
                        )
                    )
                );
                return null;
            }
        }

        let path = u?.pathname;

        if (path && String(path).startsWith('/links')) path = path.replace('/links', '');

        const searchParams = new URLSearchParams(url);

        return this.processPath(path, pathHandlers, this.paramsToObject(searchParams.entries()));
    }
}

export const DynamicLinks = new DynamicLinksService();
