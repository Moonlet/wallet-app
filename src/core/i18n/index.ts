export * from './translation/translate-component';
export * from './translation/translate';
