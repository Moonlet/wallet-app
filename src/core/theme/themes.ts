export enum Theme {
    dark = 'dark',
    light = 'light'
}
