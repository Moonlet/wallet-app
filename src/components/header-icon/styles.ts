import { StyleSheet } from 'react-native';
import { ITheme } from '../../core/theme/itheme';
import { BASE_DIMENSION } from '../../styles/dimensions';
import { normalize } from '../../library';

export default (theme: ITheme) =>
    StyleSheet.create({
        container: {
            marginHorizontal: normalize(BASE_DIMENSION * 2)
        },
        icon: {
            color: theme.colors.accent
        }
    });
