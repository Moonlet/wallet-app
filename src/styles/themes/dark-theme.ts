import { ITheme } from '../../core/theme/itheme';
import { COLORS } from '../colors';

export const darkTheme: ITheme = {
    dark: true,
    shadowGradient: ['#00000000', '#000000CC'],

    fontSize: {
        small: 12,
        regular: 16,
        large: 34
    },

    colors: {
        primary: COLORS.GRAY,
        accent: COLORS.AQUA,
        text: COLORS.WHITE,
        textDarker: COLORS.LIGHT_GRAY,
        positive: COLORS.HOT_GREEN,
        negative: COLORS.HOT_RED,

        cardBackground: COLORS.DARK_GRAY,
        appBackground: COLORS.COD_GRAY,
        settingsDivider: COLORS.BLACK
    }
};
