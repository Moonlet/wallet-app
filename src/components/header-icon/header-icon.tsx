import React from 'react';
import { View } from 'react-native';
import stylesProvider from './styles';
import { withTheme } from '../../core/theme/with-theme';
import { Icon } from '../icon';
import { normalize } from '../../styles/dimensions';

export interface IProps {
    name?: string;
    styles: ReturnType<typeof stylesProvider>;
}

export const HeaderIconComponent = (props: IProps) => (
    <View style={props.styles.container}>
        <Icon name={props.name || 'saturn-icon'} size={normalize(28)} style={props.styles.icon} />
    </View>
);

export const HeaderIcon = withTheme(stylesProvider)(HeaderIconComponent);
