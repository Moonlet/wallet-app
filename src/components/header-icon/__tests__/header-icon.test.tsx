import React from 'react';
import { HeaderIconComponent, IProps } from '../header-icon';
import { darkTheme } from '../../../styles/themes/dark-theme';
import styleProvider from '../styles';

import { shallow } from 'enzyme';
import { loadTranslations } from '../../../core/i18n';

const props: IProps = {
    styles: styleProvider(darkTheme)
};

describe('header go back button', () => {
    beforeAll(async () => loadTranslations('en'));

    it('renders correctly', () => {
        const wrapper = shallow(<HeaderIconComponent {...props} />);
        expect(wrapper.debug()).toMatchSnapshot();
    });
});
