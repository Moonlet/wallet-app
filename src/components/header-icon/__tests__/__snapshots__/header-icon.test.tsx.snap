// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header go back button renders correctly 1`] = `
"<View style={{...}}>
  <Icon name=\\"saturn-icon\\" size={28} style={{...}} allowFontScaling={false} />
</View>"
`;
